File: string/859.py
class Solution:
    def buddyStrings(self, A: str, B: str) -> bool:
        if not A or not B or len(A) != len(B):
            return False
        
        index = 0
        flag = 0
        buddy = []
        
        if A == B:
            if len(A) < 2:
                return False
            temp = set()
            for c in A:
                if c in temp:
                    return True
                temp.add(c)
            return False
        else:
            while index < len(A):
                while index < len(A) and A[index] == B[index]:
                    index += 1    

                if index == len(A) - 1 and A[index] == B[index]:
                    break
                buddy.append(index)
                index += 1    

            if len(buddy) == 2:
                if A[buddy[0]] == B[buddy[1]] and A[buddy[1]] == B[buddy[0]]:
                    return True

            return False
            
            

File: string/859.cc
class Solution {
public:
    bool buddyStrings(string A, string B) {
        if (A.size() != B.size())
            return false;
        if (A == B) {
            std::unordered_set<char> set_;
            for (auto&& c : A) {
                auto iter = set_.find(c);
                if (iter != set_.end()) {
                    return true;
                } else {
                    set_.emplace(c);
                }
            }
            return false;
        } else {    
            std::vector<size_t> list_;
            int diff = 0;
            char pre;
            for(size_t i = 0; i < A.size(); i++) {
                if (i == 0) {
                    pre = A[i];
                } else {
                    if (pre != A[i]) {
                        diff++;
                        pre = A[i];
                    }
                }
                if (A[i] != B[i]) {
                    list_.emplace_back(i);
                }
            }
            if (A == B && diff == 0)
                return true;
            if (list_.size() != 2)
                return false;
            if (A[list_[0]] == B[list_[1]] && A[list_[1]] == B[list_[0]])
                return true;
            return false;
        }
    }
};
